ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncBlockHandlerTest.java

```
@@ -0,0 +1,116 @@
+
+  @SuppressWarnings("unchecked")
+  private final BlockImporter<Void> blockImporter = mock(BlockImporter.class);
+
+  private final MutableBlockchain blockchain = mock(MutableBlockchain.class);
+  private final WorldStateArchive worldStateArchive = mock(WorldStateArchive.class);
+  private final ProtocolContext<Void> protocolContext =
+      new ProtocolContext<>(blockchain, worldStateArchive, null);
+  private final EthContext ethContext =
+      new EthContext(
+          PROTOCOL_NAME,
+          new EthPeers(PROTOCOL_NAME),
+          new EthMessages(),
+          new DeterministicEthScheduler());
+  private final LabelledMetric<OperationTimer> ethTasksTimer = NO_OP_LABELLED_TIMER;
+
+  private final FastSyncBlockHandler<Void> blockHandler =
+      new FastSyncBlockHandler<>(protocolSchedule, protocolContext, ethContext, ethTasksTimer);
+  private static final BlockHeader HEADER = new BlockHeaderTestFixture().buildHeader();
+  private static final Block BLOCK = new Block(HEADER, EMPTY_BODY);
+
+  @Before
+  public void setUp() {
+    when(protocolSchedule.getByBlockNumber(anyLong())).thenReturn(protocolSpec);
+    when(protocolSpec.getBlockImporter()).thenReturn(blockImporter);
+  }
+
+  @After
+  public void checkWorldStateIsUnused() {
+    verifyZeroInteractions(worldStateArchive);
+  }
+
+  @Test
+  public void shouldFastImportBlocks() {
+    when(blockImporter.fastImportBlock(protocolContext, BLOCK, emptyList(), LIGHT_SKIP_DETACHED))
+        .thenReturn(true);
+    final List<BlockWithReceipts> blocksWithReceipts =
+        singletonList(new BlockWithReceipts(BLOCK, emptyList()));
+
+    final CompletableFuture<List<BlockWithReceipts>> result =
+        blockHandler.validateAndImportBlocks(blocksWithReceipts);
+
+    assertThat(result).isCompleted();
+    verify(blockImporter).fastImportBlock(protocolContext, BLOCK, emptyList(), LIGHT_SKIP_DETACHED);
+  }
+
+  @Test
+  public void shouldReturnExceptionallyCompletedFutureWhenBlockImportFails() {
```

(optional) Might also be good to have a test that imports a sequence of blocks and fails early in that sequence. Then we can check that `fastImportBlock` doesn't get called for the blocks after the failure.

---

Good idea.  Done.